### ISSUE TITLE
auth-request Transaction Closing / Proper Handling

### ISSUE THREAD

**Issue by aauren** [TimWolla/haproxy-auth-request#56]
I'm going to start off by saying that I'm not a LUA expert here, so some of this is guess work.

However, we noticed while using HAProxy Ingress in Kubernetes (https://github.com/jcmoraisjr/haproxy-ingress) which integrates with a version of this plugin (https://github.com/jcmoraisjr/haproxy-ingress/blob/master/rootfs/etc/lua/auth-request.lua) that we were seeing a long lag time on successful authentication responses. Authentication failures would come through relatively quickly, but authentication successes were taking about 10 seconds to come through.

This seemed like we were running into some sort of timeout and after doing a bit of debugging, one of the developers on our team noticed that he was able to circumvent this 10 second lag by forcibly terminating the server connection via adding something like the following to the authentication server:
```
  allow(response) {
    response.status = 200;
    response.end();
    return response.connection.end();
  }
```

It seemed like whenever the server didn't terminate the connection forcibly, the client wouldn't terminate it on its own without waiting for a while. To me this sounded like the client may not have been closing the response after reading it.

After looking at the code, I can see that failures are handled differently than successes in a few ways. Specifically, when a request results in a failure it looks like the code calls `send_response()` https://github.com/TimWolla/haproxy-auth-request/blob/main/auth-request.lua#L96-L112 which culminates in a `txn:done(reply)` call.

However, it doesn't seem like successful responses ever call `txn:done(reply)`. From the documentation (https://www.arpalert.org/src/haproxy-lua-api/2.7/index.html#TXN.done), unless I miss my mark it would seem like the intent is to call this whenever processing of the transaction is complete.

Should this be added to all responses instead of just errors?

**Comment by TimWolla** [comment]
@aauren Apologies for only getting back to you just now. I've seen the notification, wasn't able to handle it and subsequently forgot.



> Should this be added to all responses instead of just errors?

No. This is for responding to the HAProxy client from HAProxy itself, instead of passing the request along to the backend. If auth-request would use that, you would be unable to access your backends.

--------------

auth-request uses the [haproxy-lua-http](https://github.com/haproxytech/haproxy-lua-http) library under the hood. From what I see it sends a `connection: close` header (preventing keep-alives from keeping the connection open) and also appears to call `socket:close()` where appropriate.

Do you have any additional information to share, e.g. packet captures between HAProxy and your authentication server?

**Comment by aauren** [comment]
@TimWolla no worries, I always feel only grateful for responses from active project maintainers. I appreciate any help you're willing to lend!

> No. This is for responding to the HAProxy client from HAProxy itself, instead of passing the request along to the backend. If auth-request would use that, you would be unable to access your backends.

Ahh... I see what you mean now. That's a good distinction. I was thinking that the full request is the response from haproxy-auth-request, but this is more-or-less a side request, and if it's successful, we still have to proxy traffic to the backend. Thanks for explaining that.

> Do you have any additional information to share, e.g. packet captures between HAProxy and your authentication server?

I can work on trying to get a packet capture going, that would probably be good debug information. I also noticed that haproxy-lua-http treats HEAD requests differently in the code (https://github.com/haproxytech/haproxy-lua-http/blob/master/http.lua#L758-L762), so I'm going to test that path and see if we get any different behavior.

Since we have a work-around for now, it may take me a week or two to get back with a response on this.

**Comment by TimWolla** [comment]
> I was thinking that the full request is the response from haproxy-auth-request, but this is more-or-less a side request, and if it's successful, we still have to proxy traffic to the backend.

Yes, exactly. auth-request takes the headers of the original client request and sends them to the authentication backend, while the original request waits for the backend's response.

> Since we have a work-around for now, it may take me a week or two to get back with a response on this.

Sure no problem. Also feel free to ping me if I don't get back to you within a week or so, because that likely means I missed the notification :smiley: 

Any other pointers (e.g. a simple standalone reproducer based on docker-compose) would also work for me, by allowing me to debug this myself.

**Comment by aauren** [comment]
@TimWolla Thanks for all of your help with this.

I've been trying to put together a small reproduction that I could share with you, but I don't seem to be able to reproduce what we're seeing with our Node app in any other language. So I think at the end of the day it must have something to do with the way that Node and possibly the frameworks that we're using treat the request.

FWIW we've discovered a couple of work-arounds that work for us that I figured I would leave on this issue in case someone else ever happens upon this issue:
* Forcibly closing the request after writing the response seems to fix the issue and causes HAProxy to return the result immediately
* Only writing a response code with no response body also seems to fix this issue for us (i.e. 200 with no response body at all)
* Converting the request to a HEAD request in HAProxy using the annotation: `auth-method: HEAD` also fixes the issue for us. Best guess is that this is because the upstream http.lua library handles HEAD requests differently: https://github.com/haproxytech/haproxy-lua-http/blob/master/http.lua#L758-L762

I'm going to close out this issue for now though because it doesn't seem to be easily reproducible for me which means that this is likely an issue with our client application rather than HAProxy or any of the libraries involved.
